livox_ros2_driver/livox_ros2_driver/lddc.cpp

```diff
@@ -712,433 +702,385 @@ void Lddc::initializeDiagnostics()
+               std::bind(&Lddc::checkVoltage, this, std::placeholders::_1,
+                         broadcast_code));
+
+  updater_.add(fmt::format("livox_motor_status-{}", broadcast_code),
+               std::bind(&Lddc::checkMotor, this, std::placeholders::_1,
+                         broadcast_code));
+
+  updater_.add(fmt::format("livox_optical_window-{}", broadcast_code),
+               std::bind(&Lddc::checkDirty, this, std::placeholders::_1,
+                         broadcast_code));
+
+  updater_.add(fmt::format("livox_firmware_status-{}", broadcast_code),
+               std::bind(&Lddc::checkFirmware, this, std::placeholders::_1,
+                         broadcast_code));
+
+  if (check_pps_signal_) {
+    updater_.add(fmt::format("livox_pps_signal-{}", broadcast_code),
+                 std::bind(&Lddc::checkPPSSignal, this, std::placeholders::_1,
+                           broadcast_code));
+  }
+
+  updater_.add(fmt::format("livox_service_life-{}", broadcast_code),
+               std::bind(&Lddc::checkServiceLife, this, std::placeholders::_1,
+                         broadcast_code));
+
+  updater_.add(
+      fmt::format("livox_fan_status-{}", broadcast_code),
+      std::bind(&Lddc::checkFan, this, std::placeholders::_1, broadcast_code));
+
+  updater_.add(fmt::format("livox_time_sync-{}", broadcast_code),
+               std::bind(&Lddc::checkTimeSync, this, std::placeholders::_1,
+                         broadcast_code));
+
+  updater_.add(fmt::format("livox_connection-{}", broadcast_code),
+               std::bind(&Lddc::checkConnection, this, std::placeholders::_1,
+                         broadcast_code));
+}
+
 void Lddc::onDiagnosticsTimer()
 {
   lidar_count_ = 0;
   for (uint32_t i = 0; i < kMaxSourceLidar; ++i) {
     if (lds_->lidars_[i].handle != kMaxSourceLidar) ++lidar_count_;
   }
+
+  if (connected_lidar_count_ != lds_->connected_lidars_.size()) {
+    connected_lidar_count_ = lds_->connected_lidars_.size();
+    for (const auto &lidar : lds_->connected_lidars_) {
+      const std::string broadcast_code = lidar.first;
+      registered_code_set_.insert(broadcast_code);
+      if (registered_code_set_.count(broadcast_code) == 0) {
+        Lddc::registerDiagnosticsUpdater(broadcast_code);
+      }
+    }
+  }
+
   updater_.force_update();
 }
 
-void Lddc::checkTemperature(diagnostic_updater::DiagnosticStatusWrapper & stat)
+void Lddc::getLidarByBroadcastcode(std::pair<std::string, DeviceInfo *> &lidar,
+                                   const std::string &broadcast_code)
 {
-  if (lidar_count_ == 0) {
-    stat.summary(DiagStatus::WARN, "No LiDARs Connected");
-    return;
+  for (const auto & connected_lidar : lds_->connected_lidars_) {
+    const std::string connected_broadcast_code = connected_lidar.first;
+    if (connected_broadcast_code == broadcast_code) lidar = connected_lidar;
   }
+}
 
-  int whole_level = DiagStatus::OK;
```

Comment:
Did you remove `whole_level` because you split diagnostics per device? (Just a confirmation.)

Reply:
Yes, that's true.